typed-racket-lib/typed-racket/types/prop-ops.rkt

```diff
@@ -70,69 +71,62 @@
 ;; Returns true if the AND of the two props is equivalent to FalseProp
 (define (contradictory? p1 p2)
   (match* (p1 p2)
-    [((TypeProp: o1 t1) (TypeProp: o2 t2))
-     #:when (object-equal? o1 o2)
+    [((TypeProp: o t1) (TypeProp: o t2))
      (not (overlap? t1 t2))]
-    [((TypeProp: o1 t1) (NotTypeProp: o2 t2))
-     #:when (object-equal? o1 o2)
+    [((TypeProp: o t1) (NotTypeProp: o t2))
      (subtype t1 t2)]
-    [((NotTypeProp: o2 t2) (TypeProp: o1 t1))
-     #:when (object-equal? o1 o2)
+    [((NotTypeProp: o t2) (TypeProp: o t1))
      (subtype t1 t2)]
-    [(_ _) (or (prop-equal? p1 -ff)
-               (prop-equal? p2 -ff))]))
+    [(_ _) (or (FalseProp? p1)
+               (FalseProp? p2))]))
 
 ;; complementary: Prop? Prop? -> boolean?
 ;; Returns true if the OR of the two props is equivalent to Top
 (define (complementary? p1 p2)
   (match* (p1 p2)
-    [((TypeProp: o1 t1) (NotTypeProp: o2 t2))
-     #:when (object-equal? o1 o2)
+    [((TypeProp: o t1) (NotTypeProp: o t2))
      (subtype t2 t1)]
-    [((NotTypeProp: o2 t2) (TypeProp: o1 t1))
-     #:when (object-equal? o1 o2)
+    [((NotTypeProp: o t2) (TypeProp: o t1))
      (subtype t2 t1)]
-    [(_ _) (or (prop-equal? p1 -tt)
-               (prop-equal? p2 -tt))]))
-
-(define (name-ref=? a b)
-  (or (equal? a b)
-      (and (identifier? a)
-           (identifier? b)
-           (free-identifier=? a b))))
+    [(_ _) (or (TrueProp? p1)
+               (TrueProp? p2))]))
 
 ;; does p imply q? (but only directly/simply)
 (define (implies-atomic? p q)
   (match* (p q)
     ;; reflexivity
-    [(_ _) #:when (or (prop-equal? p q)
-                      (prop-equal? q -tt)
-                      (prop-equal? p -ff)) #t]
+    [(_ _) #:when (or (equal? p q)
+                      (TrueProp? q)
```

Comment:
We should change the order of these tests, since the first one is now expensive (and also add `(eq? p q)` at the beginning).